### ISSUE TITLE
problems with screens query parameter

### ISSUE THREAD

**Issue by loganbraywork** [phetsims/ph-scale#171]
**Test device**

Windows 10 laptop

**Operating System**

Windows 10 v.1903

**Browser**

Chrome v. 84.0.4147.89

**Problem description**

For https://github.com/phetsims/QA/issues/515

When using the multiple screens query parameter, it is possible to create a home screen where none previously existed when using the command screens=2,1. A box does come up saying screen 2 is invalid, but a home screen appears with Macro appearing twice.

**Steps to reproduce**
 
1. Enter Query parameter ?screens=2,1

**Visuals**

![2020-07-17PHScleBasicPhetIOScreensQpar](https://user-images.githubusercontent.com/54685448/87836889-ed274700-c84e-11ea-877b-01bbd3c5b0d0.gif)

Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪pH Scale: Basics‬
URL: https://phet-dev.colorado.edu/html/ph-scale-basics/1.4.0-rc.1/phet/ph-scale-basics_all_phet.html
Version: 1.4.0-rc.1 2020-07-09 22:22:32 UTC
Features missing: touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36
Language: en-US
Window: 1433x706
Pixel Ratio: 1.3406250476837158/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 30 uniform: 4095
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {}

**Comment by pixelzoom** [comment]
I see the same behavior in master with Concentration, but not with RIAW.  They are both 1-screen sims, so I'm not sure what's causing different behavior.  Perhaps its somehow related to the fact the pH Scale: Basics and Concentration are both subsets of some other sim, but I can imagine why that would be relevant.

Assigning to @chrisklus, since I believe he worked on this feature.  High priority because this sim is in RC testing.

**Comment by samreid** [comment]
In the commit, we fall back to nominal screens if there is an error, without appending to it on subsequent loops.  It's working well in my testing.

@loganbraywork can you please test on phettest?  Please run without assertions so it will not crash during startup.

@pixelzoom or @zepumph are you available to review this?  It is just a few lines of code.

**Comment by pixelzoom** [comment]
`forEach` is an inappropriate iterator, and this clearly has not been tested well.  For example, running ph-scale-basicis with `?screens=3,4,5`, I see the same error message 3x with totally bogus values:

<img width="720" alt="screenshot_433" src="https://user-images.githubusercontent.com/3046552/88087478-35dc4a00-cb46-11ea-966c-81ba894468ee.png">

Please use a `for` loop that is exited when `allScreensValid` becomes false.  I.e.:

```
let allScreensValid = true;
for ( let i = 0; i < screensQueryParameter.length && allScreensValid; i++ ) {
   ...
}
```

**Comment by pixelzoom** [comment]
Self assigning. I'll fix this.

**Comment by pixelzoom** [comment]
@samreid please review the above commits. 

Tested cases:

// These will start without error
ph-scale-basics_en.html?screens=1
ph-scale_en.html?screens=1,2,3
ph-scale_en.html?screens=2
ph-scale_en.html?screens=2,3

// These contain errors, display warning dialog, and revert to default.
ph-scale-basics_en.html?screens=2,1
ph-scale_en.html?screens=1,4

**Comment by pixelzoom** [comment]
This issue also points to insufficient unit tests in selectScreenTests.js.  I don't have time to address that, but it should be addressed.

**Comment by pixelzoom** [comment]
I've discovered more problems with the `screens` query parameter.  So I did what I should have done to begin with - created general issue phetsims/joist#644.   

Marking this issue as "on hold" until phetsims/joist#644 is addressed.  Then fixes will need to be cherry picked to 1.4 branches of ph-scale and ph-scale-basics.

**Comment by pixelzoom** [comment]
Changing the scope and title of this issue, so that all problems identified in phetsims/joist#644 get regression tested in the context of pH Scale.

**Comment by pixelzoom** [comment]
phetsims/joist#644 has been closed. Now blocked by https://github.com/phetsims/query-string-machine/issues/42.

**Comment by pixelzoom** [comment]
Here's the first set of cherry-picks needed for joist from https://github.com/phetsims/joist/issues/644:

```
3cd429cccc852148956a00b053a1ae790c9c4dba
c29f824fc60c6ff9efbbd540d17d5764fe56e87e
e12b7b2f00fe9a1e540032e9d5ecb1b1f8664721
bde554917ede4088394bd8023cea6bce843499d2
```

**Comment by pixelzoom** [comment]
phetsims/joist#644 is closed and changes have been cherry-picked.

phetsims/query-string-machine#42 is still blocking.

**Comment by pixelzoom** [comment]
phetsims/query-string-machine#42 is resolved and has been cherry-picked. 

Ready for testing in next RC.

**Comment by pixelzoom** [comment]
To verify in ph-scale 1.4.0-rc.2, run the sim with these query parameters and verify that you get an "Invalid Query Parameters" dialog that correctly identifies the bad query parameter and its value:

* `?screens=4,1`
* `?screens=-1`
* `?screens=Macro,Micro`

Also check a few valid values, like:

* `?screens=2`
* `?screens=3,2`

**Comment by KatieWoe** [comment]
No basics to test, but it looks ok on Win 10 Chrome in 1.5.0-dev.1. @pixelzoom if that is enough, feel free to close

**Comment by brooklynlash** [comment]
On Chromebook, I only get an invalid query parameter screen with the Macro, Micro, and the valid ones work. The other two (-1, 4,1) are stuck on the phet loading screen and never load.

**Comment by pixelzoom** [comment]
@brooklynlash I'm not sure that I understand your report.  Are you referring to the tests in https://github.com/phetsims/ph-scale/issues/171#issuecomment-668840604, and saying that you're getting these results? 

* ?screens=4,1 - "Invalid Query Parameters" dialog displayed, sim stuck on Loading screen 
* ?screens=-1 - "Invalid Query Parameters" dialog displayed, sim stuck on Loading screen 
* `?screens=Macro,Micro` - "Invalid Query Parameters" dialog displayed
* `?screens=2` - sim starts as expected
* `?screens=3,2` - sim starts as expected

If so...  Are you running with any other query parameters, or just `?screens`?

**Comment by brooklynlash** [comment]
Yes, this is what I meant, and that was the only query parameter running. I can test also on Win10 and see if there is a discrepancy.

**Comment by brooklynlash** [comment]
Oh, I figured out the issue. Had the same issue on master on Win10, but then took out the brand=phet parameter and worked as expected.

**Comment by pixelzoom** [comment]
Great, thanks @brooklynlash. Closing.
